### PR TITLE
recursive prolongation and injection

### PR DESCRIPTION
**Pull request by alsgregory**
alteration:

- can now prolong and inject up more than one level of a hierarchy.
- offers support for 'generalised (or 'skipping') hierarchies' (see slack/general and firedrake-mlmc/generalised_hierarchy.py) by prolonging and injecting the full hierarchy instead of just the refined hierarchy.

### CODE REVIEW

File: firedrake/mg/interface.py [hunk 1]
```diff
@@ -18,20 +18,41 @@ def prolong(coarse, fine):
     hierarchy, lvl = utils.get_level(cfs)
     if hierarchy is None:
         raise RuntimeError("Coarse function not from hierarchy")
+    # for skipping hierarchies
+    if hasattr(hierarchy, '_full_hierarchy') == 1:
+        lvl = lvl * hierarchy._skip
+        hierarchy = hierarchy._full_hierarchy
```
Comment (wence-) [inline firedrake/mg/interface.py]: I don't understand how this works.  Neither of these properties are ever set in the hierarchy.

File: firedrake/mg/interface.py [hunk 2]
```diff
@@ -18,20 +18,41 @@ def prolong(coarse, fine):
...
+    # carry out recursive prolongs
+    coarser = Function(fhierarchy[lvl]).assign(coarse)
+    for j in range(slvl):
+        intermediate = Function(fhierarchy[lvl + j + 1])
+        op2.par_loop(fhierarchy._prolong_kernel,
+                     fhierarchy._cell_sets[lvl + j],
+                     intermediate.dat(op2.WRITE, fhierarchy.cell_node_map(lvl + j)[op2.i[0]]),
+                     coarser.dat(op2.READ, coarser.cell_node_map()))
+        if j < slvl - 1:
+            coarser = Function(fhierarchy[lvl + j + 1]).assign(intermediate)
```
Comment (wence-) [inline firedrake/mg/interface.py]: Don't keep on making functions and assigning to them:

```python
in = coarse
for j in range(slvl):
   if j == slvl - 1: # at the bottom
      out = fine
   else:
      out = Function(fhierarchy[lvl + j + 1])
   par_loop(..., out.dat(WRITE, ...), in.dat(READ, ..))
   in = out
```

File: firedrake/mg/interface.py [hunk 3]
```diff
@@ -91,16 +112,37 @@ def inject(fine, coarse):
...
+    # carry out recursive injections
+    finer = Function(hierarchy[flvl]).assign(fine)
+    for j in range(slvl):
+        intermediate = Function(hierarchy[flvl - j - 1])
+        op2.par_loop(hierarchy._inject_kernel, hierarchy._cell_sets[flvl - j - 1],
+                     intermediate.dat(op2.WRITE, intermediate.cell_node_map()[op2.i[0]]),
+                     finer.dat(op2.READ, hierarchy.cell_node_map(flvl - j - 1)))
+        if j < slvl - 1:
+            finer = Function(hierarchy[flvl - j - 1]).assign(intermediate)
+    coarse.assign(intermediate)
```
Comment (wence-) [inline firedrake/mg/interface.py]: Same here.

Comment (wence-) [inline firedrake/mg/interface.py]: I note you haven't done this for restriction.  Why not?

File: firedrake/mg/interface.py (hunk 1 above)
Comment (alsgregory) [inline firedrake/mg/interface.py]: this is explained above. see 'generalised hierarchies' in firedrakeproject/firedrake-mlmc and our previous discussion on slack channel. I need this functionality for that package.

File: firedrake/mg/interface.py (hunk 2 above)
Comment (alsgregory) [inline firedrake/mg/interface.py]: noted will change

File: firedrake/mg/interface.py (hunk 3 above)
Comment (alsgregory) [inline firedrake/mg/interface.py]: I have never used restriction before, and don't know how to use it correctly. Do you mind explaining then I will make the necessary changes.

File: firedrake/mg/interface.py (hunk 1 above)
Comment (wence-) [inline firedrake/mg/interface.py]: Yes, but firedrake doesn't know anything about that package up to now.  Now it needs to.  You should add the ability to make skipping hierarchies directly in firedrake.

File: firedrake/mg/interface.py (hunk 3 above)
Comment (wence-) [inline firedrake/mg/interface.py]: It's exactly the same as the other two (modulo the addition of the weights).

File: firedrake/mg/interface.py (hunk 1 above)
Comment (alsgregory) [inline firedrake/mg/interface.py]: Cool. I'll make the transfer of the functionality to firedrake/mg. Should this be in a separate pull request or the same as this?

Comment (wence-) [inline firedrake/mg/interface.py]: Please fix this one.

**Comment by alsgregory** [comment]
It does, but this would be M = 6. The 'skip' parameter is (M / 2). M is how many finer sub cells are in a coarse cell.

**Comment by wence-** [comment]
OK.  I would prefer something like:

```
Hierarchy(mesh, levels, refinements_per_level)
```

So `refinements_per_level = 1` corresponds to the normal hierarchy, etc...

**Comment by wence-** [comment]
You could optionally change `refinements_per_level` to be a list of length `levels` (but I'm not fussed).

File: firedrake/mg/generalised_hierarchy.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        # 'skip parameter'
+        self.skip = (M / 2)
+
+        # carry the full mesh hierarchy
+        self._full_hierarchy = MeshHierarchy(mesh, levels * self.skip)
+
+        # carry the generalised mesh hierarchy
+        self._hierarchy = self._full_hierarchy[::self.skip]
+
+        super(GeneralisedMeshHierarchy, self).__init__()
```
Comment (wence-) [inline firedrake/mg/generalised_hierarchy.py]: I'm pretty sure that you don't need a new set of classes for this at all.  I think if you just do the following in `MeshHierarchy`:

```python
def __init__(self, mesh, levels, refinements_per_level=1):
   ...
   hierarchy = all_the_refinements
   self._full_hierarchy = hierarchy
   self._hierarchy = hierarchy[::refinements_per_level]
   self.refinements_per_level = refinements_per_level
```

Then everything just falls out.  Similarly in functionspacehierarchy and friends.

**Comment by alsgregory** [comment]
Ok, I will change the argument to refinements_per_lvl. However I will keep the attribute M in to help with my multilevel package.

Here, M = (refinements_per_lvl * 2)

**Comment by wence-** [comment]
No.  I would rather it weren't there at all.  At the very least it should have a meaningful name.

File: firedrake/mg/interface.py [hunk 5]
```diff
@@ -18,20 +18,44 @@ def prolong(coarse, fine):
...
     if isinstance(hierarchy, firedrake.MixedFunctionSpaceHierarchy):
         for c, f in zip(coarse.split(), fine.split()):
             prolong(c, f)
         return
-    op2.par_loop(hierarchy._prolong_kernel,
-                 hierarchy._cell_sets[lvl],
-                 fine.dat(op2.WRITE, hierarchy.cell_node_map(lvl)[op2.i[0]]),
-                 coarse.dat(op2.READ, coarse.cell_node_map()))
+    # carry out recursive prolongs
+    coarser = Function(fhierarchy[lvl]).assign(coarse)
```
Comment (wence-) [inline firedrake/mg/interface.py]: Don't build a new function here, just write `coarser = coarse`.

**Comment by alsgregory** [comment]
No problem - it goes :)

File: tests/multigrid/test_generalised_hierarchy.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    except ValueError:
+        a = 1
+
+    assert a == 1
+
+    L = 3
+    M = 4
+    GMH = GeneralisedMeshHierarchy(mesh, L, M)
+
+    assert len(GMH) == L + 1
```
Comment (wence-) [inline tests/multigrid/test_generalised_hierarchy.py]: Code style.  Too much whitespace.  `except` should stick to the `try` block:

```python
try:
   foo = bar()
except ValueError:
   ...
```

File: tests/multigrid/test_generalised_hierarchy.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    assert GMH.M == 2
+
+    a = 0
+    try:
+        GMH = GeneralisedMeshHierarchy(mesh, L, 3)
+
+    except ValueError:
+        a = 1
+
+    assert a == 1
```
Comment (wence-) [inline tests/multigrid/test_generalised_hierarchy.py]: Nonsense.  Write this:

```python
with pytest.raises(ValueError):
   GeneralisedMeshHierarchy(mesh, L, 3)
```

File: tests/multigrid/test_generalised_hierarchy.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+    assert hasattr(GMH, '_full_hierarchy') == 1
+
+    GFSH = GeneralisedFunctionSpaceHierarchy(GMH, 'DG', 0)
+
+    assert hasattr(GFSH, '_full_hierarchy') == 1
+
+    F = Function(GFSH[0])
+
+    assert isinstance(get_level(F.function_space())[0], FunctionSpaceHierarchy) == 1
```
Comment (wence-) [inline tests/multigrid/test_generalised_hierarchy.py]: Wat?  `assert isinstance(..., FunctionSpaceHierarchy)`.  What is the `== 1` checking?

File: tests/multigrid/test_generalised_hierarchy.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    mesh = UnitSquareMesh(5, 5)
+    L = 3
+    M = 4
+    GMH = GeneralisedMeshHierarchy(mesh, L, M)
+
+    assert hasattr(GMH, '_full_hierarchy') == 1
+
+    GFSH = GeneralisedFunctionSpaceHierarchy(GMH, 'DG', 0)
+
+    assert hasattr(GFSH, '_full_hierarchy') == 1
```
Comment (wence-) [inline tests/multigrid/test_generalised_hierarchy.py]: As below.

File: tests/multigrid/test_generalised_hierarchy.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    L = 3
+    MH = MeshHierarchy(mesh, L)
+
+    a = 0
+
+    try:
+        GeneralisedFunctionSpaceHierarchy(MH, 'DG', 0)
+
+    except AttributeError:
+        a = 1
```
Comment (wence-) [inline tests/multigrid/test_generalised_hierarchy.py]: Nope.

```python
with pytest.raises(AttributeError):
   ...
```

File: tests/multigrid/test_generalised_hierarchy.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+    FSH = GeneralisedFunctionSpaceHierarchy(GMH, 'DG', 0)
+
+    assert len(FSH) == L + 1
+
+    lev = []
+    for i in range(L + 1):
+        F = Function(FSH[i])
+        lev.append(get_level(F.function_space())[1])
+
+    assert np.all(lev == np.linspace(0, L, L + 1)) == 1
```
Comment (wence-) [inline tests/multigrid/test_generalised_hierarchy.py]: `assert lev == range(L+1)`

File: tests/multigrid/test_generalised_hierarchy.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    FSH = GeneralisedFunctionSpaceHierarchy(GMH, 'DG', 0)
+
+    assert len(FSH) == L + 1
+
+    lev = []
+    for i in range((L * (M / 2)) + 1):
+        F = Function(FSH._full_hierarchy[i])
+        lev.append(get_level(F.function_space())[1])
+
+    assert np.all(lev == np.linspace(0, (L * (M / 2)), (L * (M / 2)) + 1)) == 1
```
Comment (wence-) [inline tests/multigrid/test_generalised_hierarchy.py]: Actually, don't both collecting the levels, just `assert get_level(...)[1] == i` in the loop.

**Comment by alsgregory** [comment]
Hopefully this is much better. We get rid of gen_hierarchies (but keep new tests of >1 refined hierarchies) and just modify the mesh.py and functionspace.py files directly

File: firedrake/mg/functionspace.py [hunk 13]
```diff
@@ -212,11 +217,14 @@ def __init__(self, mesh_hierarchy, family, degree, dim=None, name=None, vfamily=
...
                                                   dim=dim, name=name, vfamily=vfamily,
                                                   vdegree=vdegree)
-                for m in mesh_hierarchy]
+                for m in mesh_hierarchy._full_hierarchy]
         self.dim = fses[0].dim
         super(VectorFunctionSpaceHierarchy, self).__init__(mesh_hierarchy, fses)
 
 
+""" WE NEED TO MAKE THIS APPLICABLE FOR THE NEW TYPE OF FUNCTION SPACE HIERARCHIES """
+
```
Comment (wence-) [inline firedrake/mg/functionspace.py]: Wat?

File: tests/multigrid/test_invalid_transfers.py [hunk 14]
```diff
@@ -20,6 +20,7 @@ def f2(mesh_hierarchy):
     return FunctionHierarchy(V)
 
 
+@pytest.mark.xfail(reason="now can do recursive transfers")
```
Comment (wence-) [inline tests/multigrid/test_invalid_transfers.py]: This is not an expected failure now.  Please update the tests instead such that you test that this functionality does the right thing.

File: firedrake/mg/functionspace.py (hunk 13 above)
Comment (alsgregory) [inline firedrake/mg/functionspace.py]: sorry! left that in - an overnight comment!

File: firedrake/mg/mesh.py [hunk 15]
```diff
@@ -12,14 +12,19 @@
 
 
 class MeshHierarchy(object):
-    def __init__(self, m, refinement_levels, reorder=None):
+    def __init__(self, m, refinement_levels, reorder=None, refinements_per_lvl=1):
         """Build a hierarchy of meshes by uniformly refining a coarse mesh.
 
         :arg m: the coarse :func:`~.Mesh` to refine
         :arg refinement_levels: the number of levels of refinement
+        :arg refinements_per_lvl: optional number of refinements of mesh per level
```
Comment (wence-) [inline firedrake/mg/mesh.py]: Please spell this argument "refinements_per_level".

Comment (wence-) [inline firedrake/mg/mesh.py]: And the the instance attributes in the same way.

**Comment by alsgregory** [comment]
could this be looked over so it can be merged please?

**Comment by alsgregory** [comment]
Thanks. This would be awesome.

**Comment by wence-** [comment]
@alsgregory I have implemented skipping mesh hierarchies in the better-solvers branch.  Can you please try and see if that works for you.  I ported your idea but reimplemented a little so that you can use skipping hierarchies in multigrid.  If you ask for the level of something from a skipping hierarchy you get an integer level, if you access the `_unskipped_hierarchy` property and get the level from that it's a `Fraction`.  But you should just build things on the meshes you can iterate over in the `MeshHierarchy` and the grid transfer operations should "just work".  If that works for you, we will close this pull request.

**Comment by alsgregory** [comment]
Thanks for merging this functionality into better-solvers.

I will try all of this out when I can run firedrake-update (broken at the current moment) on this branch

**Comment by alsgregory** [comment]
I'm getting a strange error. For some reason prolong(a, b) doesn't work when level of `b` is more than 2 away from level of `a`.
Inject and restrict are fine.

**Comment by wence-** [comment]
I think I fixed that.

**Comment by wence-** [comment]
Arrived in #882.

**Pull request closed without merging**